### ISSUE TITLE
Big performance penalty due to repeated reordering of weights

### ISSUE THREAD

**Issue by yinghai** [intel/ideep#7]
We noticed that IDEEP spends quite a bit time doing reordering of tensors. This problem is especially obvious when input size is small (e.g. `batch_size=1`). After some profiling, I noticed that during our Caffe2 run, most of the time is spent redordering tensors, weight tensors in particular. This does not make sense and should be done only once, as suggested in the example of MKL-DNN (https://github.com/intel/mkl-dnn/blob/master/examples/simple_net.cpp#L798). Hopefully, we can resolve this issue as it's killing the performance and make it noncompetitive against the MKLML ops.  

Code of interest:
https://github.com/intel/ideep/blob/a861d8c2b2cceddb4d17b924b75bc1ce0a20eec0/include/ideep/computations.hpp#L1080-L1089

@gujinghui @jgong5

**Comment by 4pao** [comment]
@yinghai Yes, this is the problem source, didn't @gujinghui communicate with you? The fix will be out very soon. The real problem is not reorder the weights, instead using a very slow reorder. Because when we create tensors in caffe2 we do not know which one is data and which one is weights, so we marked weights with wrong format.

**Comment by 4pao** [comment]
@yinghai As you mentioned, we actually stuck at this place, because caffe2 weights passed as a "const" object. So we can't shove back a reordered weights back to weights blob. Any idea how to keep the weights of faster format?

**Comment by yinghai** [comment]
Yes, we just had a conference call. Just putting up an issue to track it's status. I'd still prefer to reorder only once. Even if it's fast reordering, I'd imagine it's still memory operations. And for inference, as the weights are fixed, it really makes sense to process the weights only once.

**Comment by 4pao** [comment]
@yinghai Of course. First, weights is const, we can't just const_cast it. Second, we do not know it is in inference. So, any modification can do the trick?

**Comment by yinghai** [comment]
@4pao If we can get the the descriptor of the expected weights at Caffe2 interface, for example, IDEEPOperator::GetInputDescriptor(int), we can actually do a pass of workspace rewriting to convert the weights. It's not beautiful but it's done only once upfront and it can unblock us.

Hacky Option 2:
Having prtimive W and primitive b as a member of the conv op. Upon contraction of the op, we check whether input order matches output order. If it matches then we do nothing. If it doesn't, we reorder it and assign tlit to W and b. And we set a flag say ordered_w = true. Then during runOnDevice, we can do
```
ideep::convolution_forward(Input(0), ordered_w?W:Input(1), ...);
```

**Comment by yinghai** [comment]
Option 2 only works for inference though and cannot be put into general use.

**Comment by 4pao** [comment]
@yinghai We thought about 2 once, but we can't know we are in inference. Caffe2 framework give no hint whether we are in inference except batch normalization.

**Comment by 4pao** [comment]
@yinghai The first one is what I prefer. Actually we have a function to tell framework what our weights format is. We just have to put it somewhere. I hope there is a systematical way to do it.

**Comment by yinghai** [comment]
Yeah I know. A guess weights rewriting is the way to go now. What we need to an interface to get the expected input descriptor of ideep ops.

**Comment by yinghai** [comment]
For option 2, you can just add an argument to the op, say `is_inference=1`. And we can do net writting to add that argument to each of the ideep ops. But it's still involves rewriting and I'd rather do option 1, at least we can save one copy of the weight memory from it.

**Comment by yinghai** [comment]
@4pao BTW, what's the fast reordering that you mentioned? Could you point me the code? Just curious.

**Comment by jgong5** [comment]
@yinghai The weight rewriting is supposed to be put inside "OptimizeForIdeep", right?

**Comment by yinghai** [comment]
@jgong5 I was thinking about prototyping some python script but yeah, this should be a good place to put the code.

**Comment by yinghai** [comment]
Headsup, after reviewing some of our code, I decided to fix it at the op side first: do it lazily and do it once. 

https://github.com/pytorch/pytorch/blob/a2a28c0ef1d9a433972fe72fa5b0b9b850ccfcaf/caffe2/mkl/operators/conv_op.cc#L54

In the code path, we will do shape check twice but I assume that it won't be an issue for now.

**Comment by yinghai** [comment]
Check https://github.com/yinghai/pytorch/pull/2 for a quick fix. With the fix, I'm able to get better numbers from IDEEP ops.

**Comment by yinghai** [comment]
BTW, I also notice that we are creating a new tensor each time for `group>1` cases:
https://github.com/intel/ideep/blob/master/include/ideep/computations.hpp#L1135-L1143

Should I be concerned? Is it doing data copy and transformation or it is just wearing the handle?

**Comment by 4pao** [comment]
Latter of course. The get data handle interface put into an init will disable allocation, hence no copy at all.

**Comment by yinghai** [comment]
Cool! But I dislike this line: https://github.com/intel/ideep/blob/master/include/ideep/computations.hpp#L1138 :D

**Comment by 4pao** [comment]
Well, I don't like it either BTW, MKL-DNN had no extra input for group. The group manifested itself in MKL-DNN's weights as an extra dimension, but the weights memory image is just the same as original one in caffe2 without any modification. I wish to expose this optimization technique as much as possible, then everything will be nature in API form.

There can be an helper like
weights_ = grouplize_weights(original)
then compute(X, weights...)

**Comment by yinghai** [comment]
BTW, can we have a function like
forward_inner_product::expected_weights_descriptor(), too? 

FC op has the same weight issue as conv but there is no convenient function to get expected weights descriptor like forward_convolution.

**Comment by 4pao** [comment]
Oh, the reorder might fire. We have fixed that. Push out later.

**Comment by jgong5** [comment]
@4pao Do you mean the reorder of the activation? We'd better have a memory pool design to avoid memory allocation each time. But since the tensor size would be small at that time for CNN, I guess the impact would be small.

**Comment by 4pao** [comment]
@jgong5 No, the reorder fired becase expected weights of FC is "oi" but we filled in "nc". I argued with MKL-DNN team to make them equal, and they regarded it should be the same but decide not to.

**Comment by yinghai** [comment]
Ok. Do I need to do the same check for FC op as I did for conv, or this will be fixed at IDEEP side?

**Comment by jgong5** [comment]
@yinghai Not needed for FC.

**Comment by yinghai** [comment]
@jgong5 but from above discussion, it looks like the transformation of weight may happen according to @4pao.

**Comment by 4pao** [comment]
@yinghai OK, although MKL-DNN assume nchw != oihw, but they do regard nc == io. We have no problem in two dimensional inner product.

**Comment by yinghai** [comment]
How about 4-dimensional one? I think it's a legit use case when you have FC after convs.

**Comment by 4pao** [comment]
We make them all 2 dimensional currently. In FC, weights are larger than data, it is quite cheaper to transform data than transform weights. And when comes to reorder data or not, it can be seen as a cache preload process, little performance can be squeezed.

**Comment by yinghai** [comment]
Fair enough. So far I have observed major performance penalty due to FC. But I'll keep an eye on that.

**Comment by yinghai** [comment]
This is is addressed in `Conv` so far.
